lib/exchange/provider/thorchain_exchange.provider.dart

```diff
@@ -0,0 +1,234 @@
+
+      return _thorChainAmountToDouble(expectedAmountOut);
+    } catch (e) {
+      print(e.toString());
+      return 0.0;
+    }
+  }
+
+  @override
+  Future<Limits> fetchLimits(
+      {required CryptoCurrency from,
+      required CryptoCurrency to,
+      required bool isFixedRateMode}) async {
+    final params = {
+      'from_asset': _normalizeCurrency(from),
+      'to_asset': _normalizeCurrency(to),
+      'amount': _doubleToThorChainString(1),
+      'affiliate': _affiliateName,
+      'affiliate_bps': _affiliateBps
+    };
+
+    final responseJSON = await _getSwapQuote(params);
+    final minAmountIn = responseJSON['recommended_min_amount_in'] as String? ?? '0.0';
+
+    final safeMinAmount = _thorChainAmountToDouble(minAmountIn) * 1.05;
+
+    return Limits(min: safeMinAmount);
+  }
+
+  @override
+  Future<Trade> createTrade({required TradeRequest request, required bool isFixedRateMode}) async {
+    String formattedToAddress = request.toAddress.startsWith('bitcoincash:')
+        ? request.toAddress.replaceFirst('bitcoincash:', '')
+        : request.toAddress;
+
+    final formattedFromAmount = double.parse(request.fromAmount);
+
+    final params = {
+      'from_asset': _normalizeCurrency(request.fromCurrency),
+      'to_asset': _normalizeCurrency(request.toCurrency),
+      'amount': _doubleToThorChainString(formattedFromAmount),
+      'destination': formattedToAddress,
+      'affiliate': _affiliateName,
+      'affiliate_bps': _affiliateBps,
+      'refund_address':
+          isRefundAddressSupported.contains(request.fromCurrency) ? request.refundAddress : '',
+    };
+
+    final responseJSON = await _getSwapQuote(params);
+
+    final inputAddress = responseJSON['inbound_address'] as String?;
+    final memo = responseJSON['memo'] as String?;
+
+    return Trade(
+        id: '',
+        from: request.fromCurrency,
+        to: request.toCurrency,
+        provider: description,
+        inputAddress: inputAddress,
+        createdAt: DateTime.now(),
+        amount: request.fromAmount,
+        state: TradeState.notFound,
+        payoutAddress: request.toAddress,
+        memo: memo);
+  }
+
+  @override
+  Future<Trade> findTradeById({required String id}) async {
+    if (id.isEmpty) throw Exception('Trade id is empty');
+    final formattedId = id.startsWith('0x') ? id.substring(2) : id;
+    final uri = Uri.parse('$_baseURL$_txInfoPath$formattedId');
+    final response = await http.get(uri);
+
+    if (response.statusCode == 404) {
+      throw Exception('Trade not found for id: $formattedId');
+    } else if (response.statusCode != 200) {
+      throw Exception('Unexpected HTTP status: ${response.statusCode}');
+    }
+
+    final responseJSON = json.decode(response.body);
+    final Map<String, dynamic> stagesJson = responseJSON['stages'] as Map<String, dynamic>;
+
+    final inboundObservedStarted = stagesJson['inbound_observed']?['started'] as bool? ?? true;
```

Comment:
shouldn't this default to false?

Reply:
yes, we can discard the case when a swap doesn't have a stage